datafusion/core/src/physical_optimizer/replace_with_order_preserving_variants.rs

```diff
@@ -247,7 +246,7 @@ pub(crate) fn replace_with_order_preserving_variants(
     // For unbounded cases, we replace with the order-preserving variant in any
     // case, as doing so helps fix the pipeline. Also replace if config allows.
     let use_order_preserving_variant = config.optimizer.prefer_existing_sort
-        || is_pipeline_breaking(requirements.plan.as_ref());
+        || requirements.plan.boundedness().is_unbounded();
```

Comment:
Based on the comment. I think we only need to replace order-preserving variant for unbounded + final since this case is problematic so we need to make it order preserved and hope it is able to process incrementally

Reply:
Let's assume there is a CoalesceBatchesExec, and its input is unbounded+incremental. Then you want that it would **not** be replaced with SPM?

Reply:
So, if the config flag is false, we only replace when the result generation is not possible, and that is the case only unbounded+final. That makes sense to me